### PR TITLE
@craigspaeth: [HOLD] Cache purchase

### PR DESCRIPTION
**Pull request by 1aurabrown**
builds on https://github.com/artsy/force/pull/475, ~merge that first~ actually this contains all the same commits as the other pr, so you could just merge this one but I thought it was easier to review two separate changes.
closes https://github.com/artsy/force/issues/465
might also have the same test failure that that pr has. i will look into it but wanted to push this.

puts the purchase form data in a cookie *when the modal is invoked* as it is the modal which triggers the refresh. If the modal is dismissed, the cookies is still set. If the modal is not invoked, the cookies is not set. The embedded sign in form does not require a page refresh before inquiry submission, so I thought it was unnecessary to set the cookie in any other cases. The cookie is expired once the inquiry is successfully submitted. 

I check the artwork id of the cached inquiry against the current artwork checkout page, because hypothetically you could abandon an inquiry after the cookie is set without actually submitting the inquiry and then attempt to purchase a _different_ work. I assumed it would be weird if the form of the second artwork would be pre-filled with data the user input and abandoned for a previous artwork. 

also closes https://github.com/artsy/force/issues/492

### CODE REVIEW

**Comment by 1aurabrown** [comment]
Here's a gif, sorry my debugger pops up at some point. The is the flow where I use the sign-up form with an email from an existing account and I am prompted to log in.

checkout

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please assign someone to your PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: apps/artwork_purchase/client/index.coffee [hunk 1]
```diff
@@ -2,57 +2,122 @@ Backbone = require 'backbone'
...
-    @inquiry.set {
-      name,
-      message,
-      artwork: @artwork.id,
-      contact_gallery: true,
-      inquiry_url: window.location.href,
-      user: @user
-    }
+    @$button.blur()
+    if @user.isLoggedIn() then @submitPurchaseForm() else @submitSignupForm()
```
Comment (craigspaeth) [inline apps/artwork_purchase/client/index.coffee]: I like how readable this is :)

File: apps/artwork_purchase/client/index.coffee [hunk 2]
```diff
@@ -2,57 +2,122 @@ Backbone = require 'backbone'
...
+      mode: 'login'
+      redirectTo: @artwork.href + '/checkout'
+      copy: copy
+    formData = _.extend artwork_id: @artwork.id, @purchaseForm.serializeForm()
+    Cookies.set 'purchase-inquiry', JSON.stringify formData
+
+# Signup
+  # Submit
+
+  submitSignupForm: (form, options)->
```
Comment (craigspaeth) [inline apps/artwork_purchase/client/index.coffee]: Nice work breaking up these methods into small and well documented/readable chunks!

Comment (1aurabrown) [inline apps/artwork_purchase/client/index.coffee]: I tried!

**Comment by alloy** [comment]
@1aurabrown That gif isn’t loading for me 😢

File: apps/artwork_purchase/client/index.coffee [hunk 3]
```diff
@@ -2,57 +2,122 @@ Backbone = require 'backbone'
...
+      success: @purchaseSuccess
+      error: @purchaseError
+
+  #Callbacks
+
+  purchaseSuccess: =>
+    Cookies.expire 'purchase-inquiry'
+    window.location = @artwork.href + '/thank-you'
+
+  purchaseError: =>
```
Comment (craigspaeth) [inline apps/artwork_purchase/client/index.coffee]: 👍

File: apps/artwork_purchase/client/purchase_form.coffee [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+      message,
+      artwork: @artwork.id,
+      contact_gallery: true,
+      inquiry_url: window.location.href,
+      user: @user
+    }
+
+    @inquiry.save null,
+      success: success
+      error: (model, response, options) =>
```
Comment (craigspaeth) [inline apps/artwork_purchase/client/purchase_form.coffee]: 👌

File: apps/artwork_purchase/client/index.coffee [hunk 5]
```diff
@@ -2,57 +2,122 @@ Backbone = require 'backbone'
...
-  _.extend @prototype, Form
+class PurchaseView extends Backbone.View
 
-  initialize: ({ @artwork, @user }) ->
-    @inquiry = new ArtworkInquiry
+  initialize: ({ @artwork }) ->
+    @user = User.instantiate()
+    @$button = @$ '.js-ap-summary-submit'
+
+    @purchaseForm = new PurchaseForm
```
Comment (craigspaeth) [inline apps/artwork_purchase/client/index.coffee]: Nice composition of views here.

File: apps/artwork_purchase/routes.coffee [hunk 6]
```diff
@@ -27,23 +29,35 @@ query = """
...
   return if metaphysics.debug req, res, send
   # purchaseFlow = res.locals.sd.PURCHASE_FLOW is 'purchase'
-  purchaseFlow = (res.locals.sd.NODE_ENV is 'development' or res.locals.sd.NODE_ENV is 'staging')
+  purchaseFlow = res.locals.sd.NODE_ENV is 'development' or res.locals.sd.NODE_ENV is 'staging'
   return res.redirect "/artwork/#{req.params.id}" if not purchaseFlow
+  send = query: query, variables: req.params
   metaphysics send
     .then ({ artwork }) ->
+      cookie = req.cookies['purchase-inquiry']
+      cachedData = JSON.parse cookie if cookie
```
Comment (craigspaeth) [inline apps/artwork_purchase/routes.coffee]: 👍

File: apps/artwork_purchase/routes.coffee [hunk 7]
```diff
@@ -27,23 +29,35 @@ query = """
...
 @thankYou = (req, res, next) ->
+  return if metaphysics.debug req, res, send
   # purchaseFlow = res.locals.sd.PURCHASE_FLOW is 'purchase'
   purchaseFlow = res.locals.sd.NODE_ENV is 'development' or res.locals.sd.NODE_ENV is 'staging'
   return res.redirect "/artwork/#{req.params.id}" if not purchaseFlow
-  res.redirect "/artwork/#{req.params.id}/checkout"
+  send = query: query, variables: req.params
+  metaphysics send
+    .then ({ artwork }) ->
+      return res.redirect "/artwork/#{req.params.id}" if not artwork.is_purchasable
```
Comment (craigspaeth) [inline apps/artwork_purchase/routes.coffee]: 👍

**Comment by craigspaeth** [comment]
👏  well done! This was quite a conglomerate of stateful views/flows to piece together—it doesn't get much tricky than auth and artwork inquiring.

At this point, I _would_ like to see some test boilerplate before merging. I know this is an experiment so I'm not looking for robust coverage here, just some initial pending specs on templates and client-side views and a basic test for the `@index` and `@thankYou` route handlers. Whatever is pragmatic/easy, I'm mainly interested in having some groundwork to be able quickly add tests in case, for whatever reason, we need to jump on a related help channel issue—or if the experiment wins we've got a good jumping off point to make it robust.

File: apps/artwork_purchase/client/purchase_signup_form.coffee [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+sanitizeRedirect = require 'artsy-passport/sanitize-redirect'
+Mailcheck = require '../../../components/mailcheck/index.coffee'
+Cookies = require 'cookies-js'
+
+module.exports = class PurchaseSignupForm extends Backbone.View
+  _.extend @prototype, Form
+
+  initialize: ({ @user }) -> #
+
+  # initializeMailcheck: ->
```
Comment (1aurabrown) [inline apps/artwork_purchase/client/purchase_signup_form.coffee]: Oops yes, from copy pasting.

**Comment by 1aurabrown** [comment]
no prob @craigspaeth

**Comment by craigspaeth** [comment]
🙇 thanks for the tests!

**Pull request merged**